src/rgw/rgw_bucket.cc

```diff
@@ -530,32 +531,39 @@ int rgw_remove_bucket(RGWRados *store, rgw_bucket& bucket, bool delete_children)
   if (ret < 0)
     return ret;
 
-
   RGWRados::Bucket target(store, info);
   RGWRados::Bucket::List list_op(&target);
+  CephContext *cct = store->ctx();
+  int max = 1000;
 
   list_op.params.list_versions = true;
 
-  if (delete_children) {
-    int max = 1000;
+  do {
+    objs.clear();
 
-    do {
-      objs.clear();
+    ret = list_op.list_objects(max, &objs, &common_prefixes, NULL);
+    if (ret < 0)
+      return ret;
+
+    if (!objs.empty() && !delete_children) {
+      lderr(store->ctx()) << "ERROR: could not remove non-empty bucket " << bucket.name << dendl;
+      return -ENOTEMPTY;
+    }
```

Comment:
Right. But I guess such racy conditions exist at different other places too, like between https://github.com/ceph/ceph/blob/master/src/rgw/rgw_op.cc#L2667 and https://github.com/ceph/ceph/blob/master/src/rgw/rgw_op.cc#L2686. There has to be some sort of read lock while such a operation is in progress. What do you think? Can we take that as a separate issue to fix?